Shell/Builtin.cpp

```diff
@@ -0,0 +1,743 @@
+        return 1;
+
+    if (arguments.is_empty()) {
+        for (auto& alias : m_aliases)
+            printf("%s=%s\n", alias.key.characters(), alias.value.characters());
+        return 0;
+    }
+
+    bool fail = false;
+    for (auto& argument : arguments) {
+        auto parts = String { argument }.split_limit('=', 2, true);
+        if (parts.size() == 1) {
+            auto alias = m_aliases.get(parts[0]);
+            if (alias.has_value()) {
+                printf("%s=%s\n", parts[0].characters(), alias.value().characters());
+            } else {
+                fail = true;
+            }
+        } else {
+            m_aliases.set(parts[0], parts[1]);
+        }
+    }
+
+    return fail ? 1 : 0;
+}
+
+int Shell::builtin_bg(int argc, const char** argv)
+{
+    int job_id = -1;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(job_id, "Job id to run in background", "job_id", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    if (job_id == -1 && !jobs.is_empty())
+        job_id = find_last_job_id();
+
+    Job* job = nullptr;
+
+    for (auto& entry : jobs) {
+        if (entry.value->job_id() == (u64)job_id) {
+            job = entry.value;
+            break;
+        }
+    }
+    if (!job) {
+        if (job_id == -1) {
+            printf("bg: no current job\n");
+        } else {
+            printf("bg: job with id %d not found\n", job_id);
+        }
+        return 1;
+    }
+
+    job->set_running_in_background(true);
+
+    dbg() << "Resuming " << job->pid() << " (" << job->cmd() << ")";
+    printf("Resuming job %llu - %s\n", job->job_id(), job->cmd().characters());
+
+    if (killpg(job->pgid(), SIGCONT) < 0) {
+        perror("killpg");
+        return 1;
+    }
+
+    return 0;
+}
+
+int Shell::builtin_cd(int argc, const char** argv)
+{
+    const char* arg_path = nullptr;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(arg_path, "Path to change to", "path", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    String new_path;
+
+    if (!arg_path) {
+        new_path = home;
+        if (cd_history.is_empty() || cd_history.last() != home)
+            cd_history.enqueue(home);
+    } else {
+        if (cd_history.is_empty() || cd_history.last() != arg_path)
+            cd_history.enqueue(arg_path);
+        if (strcmp(argv[1], "-") == 0) {
+            char* oldpwd = getenv("OLDPWD");
+            if (oldpwd == nullptr)
+                return 1;
+            new_path = oldpwd;
+        } else if (arg_path[0] == '/') {
+            new_path = argv[1];
+        } else {
+            StringBuilder builder;
+            builder.append(cwd);
+            builder.append('/');
+            builder.append(arg_path);
+            new_path = builder.to_string();
+        }
+    }
+
+    LexicalPath lexical_path(new_path);
```

Comment:
I don't like either building a path using concatenation or using lexical path here. Let's do it the right way, aka ask the kernel to do it.

Reply:
Not sure how I'd ask the kernel to concatenate paths, but I'll remove lexical path with realpath

Reply:
> Not sure how I'd ask the kernel to concatenate paths

To ask the kernel to concatenate a path to your current directory, you have to do this very complicated dance (...ready for it? here goes):
* just use the path

:smile:

in fact, just using the path will also take care of the absolute path case as well. Normally, you wouldn't even want to `realpath()` it, but in this case I guess you do.